psl-core/src/main/java/org/linqs/psl/evaluation/statistics/RankingEvaluator.java

```diff
@@ -289,34 +292,241 @@ public double auroc() {
             double newY = (double)tp / (double)totalPositives;
             double newX = (double)fp / (double)totalNegatives;
 
-            area += 0.5 * (newX - prevX) * (newY + prevY);
+            area += 0.5 * (newX - prevX) * Math.abs(newY - prevY) + (newX - prevX) * newY;
             prevY = newY;
             prevX = newX;
         }
 
         // Add the final piece.
-        area += 0.5 * (1.0 - prevX) * (1.0 + prevY);
+        area += 0.5 * (1.0 - prevX) * Math.abs(1.0 - prevY) + (1.0 - prevX) * 1.0;
 
         return area;
     }
 
+    /**
+     * Returns area mean Average Precision.
+     * Assumes predicted GroundAtoms are hard truth values.
+     * Assumes predicted GroundAtoms are formatted as (query, rank)
+     */
+    public double meanAveragePrecision() {
+        // hashtable to hold the number of ground truth positives for each query
+        Map<String, Integer> totalPositives = new Hashtable<>();
+
+//        System.out.println(truth.toString());
+
+        for (GroundAtom atom : truth) {
+            if (atom.getValue() >= threshold) {
+                String queryId = atom.getArguments()[0].toString();
+                if (totalPositives.containsKey(queryId)) {
+                    totalPositives.put(queryId, (totalPositives.get(queryId) + 1));
+                } else {
+                    totalPositives.put(queryId, 1);
+                }
+            }
+        }
+
+//        System.out.println("total positives" + totalPositives.toString());
+//        System.out.println(predicted.toString());
+
+        // find the average precision for each query. Note that the predictions are already sorted
+        // hashtables to hold the true positives seen so far, the position, the running sum of precisions, and
+        // the average precision for each query
+        Map<String, Integer> TPSeen = new Hashtable<>();
+        Map<String, Integer> Position = new Hashtable<>();
+        Map<String, Double> SumP = new Hashtable<>();
```

Comment:
Why `Hashtable` over `HashMap`?
Unless you need synchronized access or are playing null games, `HashMap` should be preferred.